x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/Agg.java

```diff
@@ -17,10 +20,19 @@
 
     private final String id;
     private final String fieldName;
+    private final ScriptTemplate scriptTemplate;
 
-    Agg(String id, String fieldName) {
+    Agg(String id, Object fieldOrScript) {
```

Comment:
Thinking a bit more about this, a better solution might be creating a simple DTO specific for Agg that can hold either a String or a ScriptTemplate.
That one would handle the hashCode, equals, etc.
Further more maybe extend it to accept behavior by accepting a `ValuesSourceAggregationBuilder` and internally call either `field` or `script` based on its content.
```suggestion
    Agg(String id, AggTarget target) {
       this.target = target;
    }
    ...
    class AggTarget {
        private final String fieldName;
        private final ScriptTemplate script;
        
        ...
        
        ValuesSourceAggregationBuilder with(ValuesSourceAggregationBuilder aggBuilder) {
            if (field != null) {
                 aggBuilder.field(field);
             }
             else {
                 aggBuilder.script(scriptTemplate.asScript());
              }
        }
    }
```

This would improve `LeafAgg` which would improve `toBuilder` and call `AggTarget` so that subclasses would only have to create the `ValueSourceAggregationBuilder`. In fact we could just declare them as a method reference for a string:

```java
public abstract class LeafAgg extends Agg {

    LeafAgg(String id, AggTarget target) {
        super(id, target);
    }

    ValuesSourceAggregationBuilder builder() {
         return aggBuilder.apply(id()).with(aggTarget);
    }

    protected abstract Function<String, ValuesSourceAggregationBuilder> aggBuilder();
```

so that subclasses would now look like:
```java

public class CardinalityAgg extends LeafAgg {

    public CardinalityAgg(String id, AggTarget aggTarget) {
        super(id, aggTarget);
    }

    @Override
    ValuesSourceAggregationBuilder aggBuilder() {
        return AggregationBuilders::cardinality;
    }
}
```

Reply:
Thx @costin, It's cleaner this way!